src/librbd/mirror/snapshot/PromoteRequest.cc

```diff
@@ -0,0 +1,392 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/mirror/snapshot/PromoteRequest.h"
+#include "common/Timer.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/ImageState.h"
+#include "librbd/Operations.h"
+#include "librbd/Utils.h"
+#include "librbd/image/ListWatchersRequest.h"
+#include "librbd/mirror/snapshot/CreateNonPrimaryRequest.h"
+#include "librbd/mirror/snapshot/CreatePrimaryRequest.h"
+#include "librbd/mirror/snapshot/Utils.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::mirror::snapshot::PromoteRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace mirror {
+namespace snapshot {
+
+using librbd::util::create_async_context_callback;
+using librbd::util::create_context_callback;
+
+template <typename I>
+void PromoteRequest<I>::send() {
+  refresh_image();
+}
+
+template <typename I>
+void PromoteRequest<I>::refresh_image() {
+  if (!m_image_ctx->state->is_refresh_required()) {
+    create_promote_snapshot();
+    return;
+  }
+
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  auto ctx = create_context_callback<
+    PromoteRequest<I>, &PromoteRequest<I>::handle_refresh_image>(this);
+  m_image_ctx->state->refresh(ctx);
+}
+
+template <typename I>
+void PromoteRequest<I>::handle_refresh_image(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to refresh image: " << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  if (!util::can_create_primary_snapshot(m_image_ctx, false, m_force,
+                                         &m_rollback_snap_id)) {
+    if (!m_force) {
+      lderr(cct) << "cannot promote" << dendl;
+      finish(-EINVAL);
+      return;
+    }
+    ldout(cct, 0) << "force promoting without rollback to known state" << dendl;
```

Comment:
If you are force-promoting, you should always rollback if the last snapshot wasn't a non-primary mirror snapshot and it wasn't complete.

Reply:
Actually, it was for the case when the image in a state when we can't find any snapshot we could rollback to. I have no idea if it would possible to get such a state, probably after some error on user end. Should we just fail 'force promotion' in this case? The user might still want to use such an image and force mirroring. But now I think she could probably just disable/enable mirroring for such an image to clear the state?

Reply:
>  But now I think she could probably just disable/enable mirroring for such an image to clear the state?

That sounds like a good workaround for now should the case arise.